### PR TITLE
Ftr/text vectorization ext

### LINKED ISSUE #31: Load ML model into Tensorflow.js

### LINKED ISSUE #36: Serialize Tokenization process
Although the ML model can be easily serialized using Tensorflow functions, the `TextVectorization` layer we were using previously cannot be. A new method of serializing this process must be found, and it must be compatible with javascript.

### PR DESCRIPTION
**Pull request by TempixTL**
Fixes #36

### CODE REVIEW

**Comment by TempixTL** [comment]
### Metrics

- Categorical Accuracy: 83%
- Precision: 0.84
- Recall: 0.82

### Confusion Matrix on held-out set

| | Predicted Low-Calorie | Predicted Average-Calorie | Predicted High-Calorie |
| --- | --- | --- | --- |
| **Actual Low-Calorie** | 11285 | 1403 | 699 |
| **Actual Average-Calorie** | 2003 | 12097 | 2823 |
| **Actual High-Calorie** | 741 | 1786 | 21509 |

**Comment by TempixTL** [comment]
Prematurely updated model, undone with latest force-push

**Comment by TempixTL** [comment]
Updated model so that Tokenization file (`tokenizer.json`) is accurate.

### Metrics

- Categorical Accuracy: 83%
- Precision: 0.84
- Recall: 0.82

### Confusion Matrix on held-out set

| | Predicted Low-Calorie | Predicted Average-Calorie | Predicted High-Calorie |
| --- | --- | --- | --- |
| **Actual Low-Calorie** | 11096 | 1601 | 648 |
| **Actual Average-Calorie** | 1673 | 12788 | 2264 |
| **Actual High-Calorie** | 782 | 2298 | 21196 |

File: amaranth/ml/train.py [hunk 1]
```diff
@@ -53,23 +55,34 @@ def main():
...
   tokenized_corpus = corpus.map(lambda desc: desc.split(' '))
   max_corpus_length = lib.max_sequence_length(tokenized_corpus)
 
-  # Create TextVectorization layer
-  vectorize_layer = keras.layers.experimental.preprocessing.TextVectorization(
-      max_tokens=vocab_size,
-      output_sequence_length=max_corpus_length,
-  )
+  # Encode 'description' into new column 'input'
+  calorie_data['input'] = calorie_data.apply(
```
Comment (timmy-ch) [inline amaranth/ml/train.py:66]: Make sure you're also implementing
- a max_token filter (which limits only the top X number of words to become tokens)
- a out-of-vocabulary (OOV) token, which you assign for every word that isn't part of your dictionary

Comment (TempixTL) [reply amaranth/ml/train.py]: Very true, forgot about this since keras' TextVectorization layer did that for me. Fixed in https://github.com/googleinterns/amaranth/pull/35/commits/c5622901df47d6457b6cc7f5e720c58a48c6102a

File: amaranth/ml/train.py [hunk 2]
```diff
@@ -17,12 +18,16 @@
...
 RESOURCES_DIR = '../resources/'  # Project resources directory
-
+CHROME_EXT_DIR = 'amaranth-chrome-ext/'  # Chrome extension directory
 # Fraction of data that should be used for training, validation, and testing.
 # Should all sum to 1.0.
 TRAIN_FRAC = 0.6
 VALIDATION_FRAC = 0.2
 TEST_FRAC = 0.2
+# Times a token needs to appear to be in model's vocab
+MIN_TOKEN_APPEARANCE = 0
```
Comment (timmy-ch) [inline amaranth/ml/train.py:28]: to be clear, you are adding every token? how do you deal with OOV tokens then?

File: amaranth/ml/train.py [hunk 3]
```diff
@@ -52,24 +57,69 @@ def main():
...
+
+  # Only 'remember' words that appear at least 3 times
+  for word, cnt in tokenizer_cnt.items():
+    if cnt < MIN_TOKEN_APPEARANCE:
+      del tokenizer[word]
+
+  # The empty string denotes words that are OOV (out-of-vocabulary)
+  # It is equal to vocab_size because all values in the tokenizer should be less
+  # than vocab_size
+  tokenizer[''] = 0
```
Comment (timmy-ch) [inline amaranth/ml/train.py:103]: 1) if an item is actually OOV, might be worth calling it OOV i.e.
``` tokenizer['OOV'] = 0 ```
2) Just curious, but why are you assigning 0 as the value?

File: amaranth/ml/train.py [hunk 4]
```diff
@@ -52,24 +57,69 @@ def main():
...
-  )
-
-  vectorize_layer.adapt(calorie_data['description'].to_numpy())
+  # Encode 'description' into new column 'input'
+  calorie_data['tokenized'] = calorie_data.apply(
+      lambda row: text.one_hot(row['description'], vocab_size), axis=1)
+
+  # Create tokenizer to encode input text
+  tokenizer = dict()  # Dictionary mapping each unique word to a unique int
+  tokenizer_cnt = dict()  # Counts appearances of each word in tokenizer
```
Comment (timmy-ch) [inline amaranth/ml/train.py:85]: Might be worth using defaultdict(int) here. Example: https://docs.python.org/3/library/collections.html#collections.defaultdict

File: amaranth/ml/train.py (hunk 2 above)
Comment (TempixTL) [reply amaranth/ml/train.py:28]: Shoot, I set it to zero during debugging. It should be 3, I'll check that out!

File: amaranth/ml/train.py [hunk 5]
```diff
@@ -52,24 +57,69 @@ def main():
...
+  for _, row in calorie_data.iterrows():
+    for idx, word in enumerate(row['description'].split()):
+      tokenizer[word] = row['tokenized'][idx]
```
Comment (timmy-ch) [inline amaranth/ml/train.py:88]: Does this actually create a unique int per word? I'd imagine it repeats from 0-n per sentence. Also, there's a decent chance words actually repeat so you'd want to make sure that doesn't happen either?

File: amaranth/ml/train.py (hunk 3 above)
Comment (TempixTL) [reply amaranth/ml/train.py:103]: 1. Good point, will change that!
2. As my (accidentally outdated) comment states, it used to be equal to `vocab_size` but that caused issues because the model expects values strictly less than `vocab_size`. I could tell the model to expect values thru `vocab_size+1`, but Tensorflow.js currently has no way to determine what `vocab_size` is/was

File: amaranth/ml/train.py (hunk 5 above)
Comment (TempixTL) [reply amaranth/ml/train.py:88]: It does create a unique int per word. The `row['tokenized']` series is a list of ints created using [keras' one-hot-encoding function](https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/text/one_hot) before the loop starts, this loop just populates the dictionary with those values so it can be serialized to json!

File: amaranth/ml/train.py (hunk 3 above)
Comment (timmy-ch) [reply amaranth/ml/train.py:103]: Right. I think when you create the model below, you also need to make sure the embedding sizes are correct since technically you are adding one additional "word". Also, in general, it might be useful to just define your OOV token as 0; that way, the vocab_size variable doesn't apply on the JS side of the equation. 
```
keras.layers.Embedding(vocab_size+1, int((vocab_size+1)**(0.25) ...)
```

File: amaranth/ml/train.py (hunk 5 above)
Comment (timmy-ch) [reply amaranth/ml/train.py:88]: Ah okay, that sounds good. Also worth noting it seems like it generates ints = [1, n] which means you can always keep OOV=0

Comment (TempixTL) [reply amaranth/ml/train.py:88]: Good catch, that's convenient!

Comment (timmy-ch) [reply amaranth/ml/train.py:88]: Yup. Just make sure about the uniqueness guarantee? The site does say "unicity non-guaranteed"

File: amaranth/ml/train.py (hunk 2 above)
Comment (TempixTL) [reply amaranth/ml/train.py]: Fixed in https://github.com/googleinterns/amaranth/pull/35/commits/f5ffbf7e2d8fa72bffdb29d0e587e6cc3fb6115c

File: amaranth/ml/train.py (hunk 3 above)
Comment (TempixTL) [reply amaranth/ml/train.py]: Point 1 fixed in https://github.com/googleinterns/amaranth/pull/35/commits/4e02b52d0afefda12671b1113402764622c952f2.

> Right. I think when you create the model below, you also need to make sure the embedding sizes are correct since technically you are adding one additional "word".
> 
> ```
> keras.layers.Embedding(vocab_size+1, int((vocab_size+1)**(0.25) ...)
> ```

Fixed in https://github.com/googleinterns/amaranth/pull/35/commits/9eb700437b03de7bac9a812bd470f2c5fdd99838

File: amaranth/ml/train.py (hunk 4 above)
Comment (TempixTL) [reply amaranth/ml/train.py]: Changed in https://github.com/googleinterns/amaranth/pull/35/commits/09c21da0537a006070af43b3301c511759b25d81

File: amaranth/ml/train.py (hunk 5 above)
Comment (TempixTL) [reply amaranth/ml/train.py:90]: You're totally right. When I came across that I was transitioning to the TextVectorization layer, so I thought it wouldn't matter after that. But since we're back to rolling our own vectorization code that's something I'll have to fix

File: amaranth/ml/train.py [hunk 6]
```diff
@@ -82,25 +83,22 @@ def main():
 
   # Create tokenizer to encode input text
   tokenizer = dict()  # Dictionary mapping each unique word to a unique int
-  tokenizer_cnt = dict()  # Counts appearances of each word in tokenizer
+  tokenizer_cnt = defaultdict(
```
Comment (timmy-ch) [inline amaranth/ml/train.py:86]: defaultdict(int) defaults to 0. https://docs.python.org/3/library/collections.html#defaultdict-examples

Comment (TempixTL) [reply amaranth/ml/train.py:86]: Oh, even better! I'll adjust

File: amaranth/ml/train.py (hunk 5 above)
Comment (TempixTL) [reply amaranth/ml/train.py]: Fixed in https://github.com/googleinterns/amaranth/pull/35/commits/27f9a674db14f5a5caf457c7d3e06ac2481e69c9

File: amaranth/ml/train.py (hunk 6 above)
Comment (TempixTL) [reply amaranth/ml/train.py]: Updated in https://github.com/googleinterns/amaranth/pull/35/commits/20b457cc1a4bd755ab253dabffe659f1418f33ab

**Comment by TempixTL** [comment]
Updated model due to further tokenization changes.

### Metrics

- Categorical Accuracy: 83%
- Precision: 0.84
- Recall: 0.82

### Confusion Matrix on held-out set

| | Predicted Low-Calorie | Predicted Average-Calorie | Predicted High-Calorie |
| --- | --- | --- | --- |
| **Actual Low-Calorie** | 11185 | 1590 | 609 |
| **Actual Average-Calorie** | 1612 | 13053 | 2166 |
| **Actual High-Calorie** | 657 | 2471 | 21003 |

File: amaranth/ml/train.py [hunk 7]
```diff
@@ -52,26 +57,70 @@ def main():
...
-      output_sequence_length=max_corpus_length,
-  )
+  # Count appearances of each word in dataset
+  tokenizer_cnt = defaultdict(int)
+  for dish_name in calorie_data['description']:
+    for token in dish_name.split():
+      tokenizer_cnt[token] += 1
+
+  # Only 'remember' words that appear at least MIN_TOKEN_APPEARANCE times
+  keep_tokens = [
```
Comment (timmy-ch) [inline amaranth/ml/train.py:86]: nit: I know this makes things into one line, but it's honestly hard to read. Do you mind writing this as a for loop?

Comment (TempixTL) [reply amaranth/ml/train.py:86]: I see how that could be hard to read, I'll refactor!

Comment (TempixTL) [reply amaranth/ml/train.py]: Fixed in https://github.com/googleinterns/amaranth/pull/35/commits/783d9d36ecb4e226f0d1b8ee1d0bb1a187c37d10
